### PR TITLE
POS-98

### PR DESCRIPTION
**Pull request by 0xSasaPrsic**
Implementation of tests for PBFT consensus.

### CODE REVIEW

File: consensus.go [hunk 1]
```diff
@@ -600,25 +614,25 @@ func (p *Pbft) runRoundChangeState(ctx context.Context) {
...
 }
 
-func (p *Pbft) sendCommitMsg() {
-	p.gossip(MessageReq_Commit)
+func (p *Pbft) sendCommitMsg() error {
+	return p.gossip(MessageReq_Commit)
 }
 
-func (p *Pbft) gossip(typ MsgType) {
+func (p *Pbft) gossip(msgType MsgType) error {
```
Comment (ferranbt) [inline consensus.go:633]: So what should we do if there is a gossip error? What type of errors can we have in the gossip protocol? My understanding is that the Transport and Gossip function are async functions that should not fail to the user but handle any errors themselves.

Internally, a gossip transports enqueues all the possible messages from the Peer and does its best to broadcast them. I think we could even remove the error from the function altogether. Similar though applies to Sign. Thus, I think we can remove this errors from the gossip functions.

File: e2e/e2e_node_drop_test.go [hunk 2]
```diff
@@ -10,8 +13,57 @@ func TestE2E_NodeDrop(t *testing.T) {
...
+func TestE2E_NetworkChurn(t *testing.T) {
+	rand.Seed(time.Now().Unix())
+	nodeCount := 20
+	maxFaulty := nodeCount/3 - 1
+	const prefix = "ptr_"
+	c := newPBFTCluster(t, "network_churn", "ptr", nodeCount)
+	c.Start()
+	runningNodeCount := nodeCount
+	// randomly stop nodes every 3 seconds
+	executeInTimerAndWait(3*time.Second, 30*time.Second, func(_ time.Duration) {
```
Comment (ferranbt) [inline e2e/e2e_node_drop_test.go:31]: Do not make this an E2E test but a fuzz test. Create a different file called ```fuzz_network_churn_test.go``` and call the test ```TestFuzz_**``` instead.

File: e2e/framework.go [hunk 3]
```diff
@@ -188,11 +188,25 @@ func (c *cluster) WaitForHeight(num uint64, timeout time.Duration, nodes ...[]st
 				return
 			}
 		case <-timer.C:
-			c.t.Fatal("timeout")
+			if throwPanic {
```
Comment (ferranbt) [inline e2e/framework.go:191]: Why do you want to throwPanic here? What is the workflow of the tests that use this panic?

File: msg_queue_test.go [hunk 4]
```diff
@@ -46,18 +44,41 @@ func TestMsgQueue_RoundChangeState(t *testing.T) {
...
+		m = newMsgQueue()
+		m.pushMessage(mockQueueMsg("A", MessageReq_RoundChange, ViewMsg(3, 1)))
+		assert.Nil(t, m.readMessage(RoundChangeState, ViewMsg(1, 1)))
+
+		m.pushMessage(mockQueueMsg("A", MessageReq_Commit, ViewMsg(3, 1)))
+		assert.Nil(t, m.readMessage(CommitState, ViewMsg(1, 1)))
+	}
+}
+
+func Test_msgToState(t *testing.T) {
```
Comment (ferranbt) [inline msg_queue_test.go:65]: I do not find it necessary to test that it panics.

File: e2e/e2e_unreliable_network_test.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+import (
+	"fmt"
+	"math/rand"
+	"strconv"
+	"testing"
+	"time"
+)
+
+func TestE2E_Unreliable_Network(t *testing.T) {
```
Comment (ferranbt) [inline e2e/e2e_unreliable_network_test.go:11]: Is this a Fuzz or E2E test?

File: Makefile [hunk 6]
```diff
@@ -3,6 +3,6 @@ test:
 	go test -v ./...
 
 e2e:
-	cd ./e2e && go test -v ./...
+	cd ./e2e && go test -timeout=50m -v ./...
```
Comment (ferranbt) [inline Makefile:6]: If we only run the E2E tests in CI ```go test -v ./... -run TestE2e``` we should be able to remove this timeout. The Fuzz tests will not be tested on each PR but nightly most likely.

File: consensus.go [hunk 7]
```diff
@@ -378,23 +394,22 @@ func (p *Pbft) runValidateState(ctx context.Context) { // start new round
...
+  
+	timeout := p.roundTimeout(p.state.view.Round)
 
-	timeout := p.exponentialTimeout()
 	for p.getState() == ValidateState {
 		_, span := p.tracer.Start(ctx, "ValidateState")
 
 		msg, ok := p.getNextMessage(span, timeout)
-		if !ok {
+		if !ok || msg == nil {
```
Comment (ferranbt) [inline consensus.go:404]: Please revert this change. I usually have a hard time myself with weird logic operands so I avoid them as much as possible.

File: consensus_test.go [hunk 8]
```diff
@@ -11,287 +15,754 @@ import (
...
+type MockedTracer struct {
+	mock.Mock
+}
+
+func (m *MockedTracer) Start(ctx context.Context, spanName string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
+	args := m.Called(ctx, spanName, opts)
+	return args.Get(0).(context.Context), args.Get(1).(trace.Span)
+}
+
+func Test_InitializeConfigOptionsSet(t *testing.T) {
```
Comment (ferranbt) [inline consensus_test.go:38]: We do not need to do this test.

File: consensus_test.go [hunk 9]
```diff
@@ -11,287 +15,754 @@ import (
...
+
+	assert.Equal(t, timeout, m.Pbft.config.Timeout)
+	assert.Equal(t, proposalTimeout, m.Pbft.config.ProposalTimeout)
+	assert.Equal(t, mockedTracer, m.Pbft.config.Tracer)
+
+	mockedTracer.AssertNotCalled(t, "Start")
+	mockedTracer.AssertExpectations(t)
+}
+
+func TestTransition_AcceptState(t *testing.T) {
```
Comment (ferranbt) [inline consensus_test.go:73]: Please revert this change. It is fine but when you are debugging or changing parts of the consensus and want to test specific functions, it is cumbersome to have to comment out this array instead of running ```go test -v ./... -run TestAcceptState_Something```.

File: consensus_test.go [hunk 10]
```diff
@@ -11,287 +15,754 @@ import (
...
+		sc := sc
+		f := f
+		t.Run(sc, func(t *testing.T) {
+			t.Parallel()
+			f(t)
+		})
+	}
+}
+
+func TestTransition_RoundChangeState(t *testing.T) {
```
Comment (ferranbt) [inline consensus_test.go:99]: Same as in TestTransition_AcceptState.

File: consensus_test.go [hunk 11]
```diff
@@ -11,287 +15,754 @@ import (
...
+		sc := sc
+		f := f
+		t.Run(sc, func(t *testing.T) {
+			t.Parallel()
+			f(t)
+		})
+	}
+}
+
+func TestTransition_ValidateState(t *testing.T) {
```
Comment (ferranbt) [inline consensus_test.go:120]: Same as in TestTransition_AcceptState.

**Review by ferranbt — CHANGES_REQUESTED** [review]

**Comment by ferranbt** [comment]
Please revert anything that is not test updates from consensus_test.go. There are so many changes in the way tests are run + variable changes that I cannot even figure out if there are any changes to the tests.

File: e2e/e2e_node_drop_test.go (hunk 2 above)
Comment (0xSasaPrsic) [reply e2e/e2e_node_drop_test.go:31]: :+1:

File: e2e/e2e_unreliable_network_test.go (hunk 5 above)
Comment (0xSasaPrsic) [reply e2e/e2e_unreliable_network_test.go:11]: It's fuzz test, will make adjustments as for previous fuzz test.

File: consensus.go (hunk 1 above)
Comment (sprsic) [reply consensus.go:633]: sure, I will make adjustments.

File: e2e/framework.go (hunk 3 above)
Comment (igorcrevar) [reply e2e/framework.go:191]: This is added because we need somehow to detect timeout in tests. We found that it is easier to detect timeout in tests with panic/recover and we thought `WaitForHeight` is just a simple function inside e2e testing framework so introducing one more parameter is good enough solution. Otherwise the only other solution is to mock logs and catch t.Fatal("timeout")

File: consensus.go (hunk 1 above)
Comment (0xSasaPrsic) [reply consensus.go:633]: sure, I will make adjustments.

Comment (Stefan-Ethernal) [reply consensus.go]: > So what should we do if there is a gossip error? 

We had this example on our mind when been thinking through it and thought maybe it is better to return an `error` object and caller can decide how to handle it. The question in the snippet below is: is it ok here to proceed with sending `prepare` messages if for some reason `sendPreprepareMsg` failed? If the answer is affirmative, then we should discuss and agree upon error handling strategy for each use case which can happen.
```
func (p *Pbft) runAcceptState(ctx context.Context) {
...
   // send the preprepare message
   p.sendPreprepareMsg()
   // send the prepare message since we are ready to move the state
   p.sendPrepareMsg()
...
}
```
On the other hand, we don't mind with only logging out errors, but thought that returning error object would be convenient.

> What type of errors can we have in the gossip protocol?

We don't know that yet, those are implementation details of validator seal signing (`Sign` function), gossiping on transport layer (`Gossip` function) etc. Namely each function which is being invoked within a 
`func (p *Pbft) gossip(msgType MsgType)` function.

**Comment by 0xSasaPrsic** [comment]
> Please revert anything that is not test updates from consensus_test.go. There are so many changes in the way tests are run + variable changes that I cannot even figure out if there are any changes to the tests.

Hey @ferranbt I made some changes, but there are some questions left in the comments, could you please take a look?

File: e2e/e2e_node_drop_test.go (hunk 2 above)
Comment (ferranbt) [reply e2e/e2e_node_drop_test.go]: Yes, everything that is not "deterministic" and has some random scenarios in the test itself should not be part of e2e_, all of those go to fuzz_xxx_test.go but keep on the same package.

File: e2e/e2e_unreliable_network_test.go (hunk 5 above)
Comment (ferranbt) [reply e2e/e2e_unreliable_network_test.go]: Move to fuzz_xxx_test.go and make it a fuzz test.

File: consensus.go (hunk 1 above)
Comment (ferranbt) [reply consensus.go]: What are you going to do in the PBFT state machine if sending Preprepare fails?

Comment (Stefan-Ethernal) [reply consensus.go]: I was thinking maybe interrupting state machine loop and propagate an error back to the client immediately and it could decide about some retry logic, if it is important to send `prepare` message only after successful `pre-prepare` message sending. 
However, I got your point as well and seems this is not the case:
>  ...a gossip transports enqueues all the possible messages from the Peer and does its best to broadcast them.

which suggests that a node is doing its best to deliver all the messages to the peers and there is anyway `RoundChange` state when some essential condition is unsatisfied.

Just for the note, we have already reverted this change, just wanted to discuss around it a little bit more. :slightly_smiling_face:

File: e2e/e2e_unreliable_network_test.go (hunk 5 above)
Comment (0xSasaPrsic) [reply e2e/e2e_unreliable_network_test.go]: Fuzz tests moved to separate fail with appropriate test function name.

**Comment by sprsic** [comment]
Closing this PR as it will be created in smaller chunks.

**Comment by 0xSasaPrsic** [comment]
Closing this PR as it is going to be created in several smaller chunks.
